lib/events/azsessions/azsessions.go

```diff
@@ -0,0 +1,518 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package azsessions
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net/url"
+	"path"
+	"strconv"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/streaming"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/session"
+)
+
+const (
+	sessionContainerName    = "session"
+	inprogressContainerName = "inprogress"
+
+	// uploadMarkerPrefix is the prefix for upload markers, stored at `upload/<session ID>/<upload ID>`.
+	uploadMarkerPrefix = "upload/"
+	// uploadMarkerFmt is the format string for upload markers, stored at `upload/<session ID>/<upload ID>`.
+	uploadMarkerFmt = "upload/%v/%v"
+
+	// partFmt is the format string for upload parts, stored at `part/<session ID>/<upload ID>/<part number>`.
+	partFmt = "part/%v/%v/%v"
```

Comment:
Up to you, but I'm generally not a fan of constants for format strings, because then it's not clear at the call site whether or not the correct number of varargs are passed in.

Reply:
Turned those into functions instead. I really just wanted a place where you could see all the types of path we use, rather than have them all over the place.